test/test_framework.py

```diff
@@ -843,6 +879,27 @@ class SomeObject(Object):
                 self.assertEqual(a, old_a)
                 self.assertEqual(b, old_b)
 
+    def test_dirty_state(self):
+        class SomeObject(Object):
+            state = StoredState()
+
+        framework = self.create_framework()
+        obj = SomeObject(framework, '1')
+        obj.state.v = 42
```

Comment:
If each one of the dirty flag changes in the code are individually commented out, and the test is run again, I suspect most are still uncovered.

Reply:
Agreed, I reworked the test significantly.

The only place I could find that does not affect test results now is `self.dirty = False` in `on_commit`. This is because it's also done in `restore`.

Also:
* moved the list of test cases for mutable types to the `TestStoredState`'s `__init__` method to share it across `test_mutable_types` and `test_dirty_state`;
  * also converted the list to a tuple to avoid inadvertent modification of the test table in any of the test cases;
* added a missing `list[i] = x` test by calling __setitem__ directly (cannot use assignment in lambdas);
* modified `test_mutable_types` to commit after attribute assignment because it did not catch any errors in setting the dirty flag on data structure modification (`__setitem__` on StoredStateData itself made the state dirty).